### ISSUE TITLE
Request HEFER production for one or multiple SBIs

### ISSUE THREAD

**Issue by CrispinF** [exeGesIS-SDM/ELM#6]
Workflow for RPA user to initiate HEFER production for one or multiple SBIs. 
Permit pasting a list of SBIs into a text area, with some range of permissible formats (e.g. each on a line, or comma separated). Possibly check and validate the entered SBIs to check they are viable (regex?). Or maybe no need for this if each is going to be used to fetch parcels from Defra web service, as that itself will be validation.

**Comment by CrispinF** [comment]
@sarahpoppy again, terminology decisions needed.

**Comment by sarahpoppy** [comment]
Again, can't the coding element use generic language, with terminology to be applied cosmetically?

**Comment by CrispinF** [comment]
@joerichardsESDM I'd suggest a templated page similar to what Jim did for the "retrieve HEFER", allowing editable texts around a functional control. The data entry needs only to accept a valid SBI and insert it into the table with appropriate metadata.

**Comment by CrispinF** [comment]
@JimKerslake we'll need a quick conflab to show Joe how this works and decide what editable areas and controls it needs.

**Comment by CrispinF** [comment]
@JimKerslake to develop this.

**Comment by CrispinF** [comment]
Suggested URL: /hefer-request
Editable main title, text block (html), control label (as for Download page).
Text control - single line, with regex validation (pattern to be confirmed but something like "SBI" & 10 digits). Perhaps permit trailing spaces, and trim them on database entry.
"Submit" button.

If fail validation, client side: "The SBI does not appear to be valid; please enter a valid SBI comprising the letter SBI and ten digits, e.g. SBI 1234598764" (or whatever).

If validation succeeds, post back and insert the SBI plus metadata (user, when) into table TBC.

Feedback on submission: redirect to /hefer-request-submitted
with editable main title and text block (html).

Incorrect SBIs will be detected further down the process, not here.

**Comment by CrispinF** [comment]
@joerichardsESDM can you confirm the pattern for valid SBIs please?
Also, could you add here some info about what table to push the SBI into, and what additional fields to populate.

**Comment by joerichardsESDM** [comment]
The regex we use in SHINE is `/^\d{9}[a-zA-Z]{0,2}$/` but of course SHINE allows a suffix and this won't, so I think a simple `/^\d{9}$/` will suffice

**Comment by joerichardsESDM** [comment]
Creating the PackRequest should be through Entity Framework. 

The parent object to create is **PackRequest** with appropriate values in its **Sbi**, **CreatedDateTime** and **CreatedByUser** properties, the latter being the Cloudscribe UserGuid as a string.

There should be a child **ParcelLoad** object with **CreatedDateTime** set and **Status** set to 1 (you can use enum ParcelLoadStatus.NotStarted)

I'd suggest creating a function in Elm.Data.Services.PackRequests.PackRequestsService to do all of this.

**Comment by CrispinF** [comment]
@joerichardsESDM ok I think you are saying that the "SBI" prefix is not part of the identifier, and the applicants & RPA only use the 9 digits? Correct?

**Comment by JimKerslake** [comment]
Necessary first steps (just for reference)
- Create text pack xml definitions and c# config class, for the editable page text elements, and register into config provider
- Check config provider admin UI, and create an Override for this text pack class
- Manually create view and import ConfigResolver and editableTextParser
- Render text out onto view, using the EditableText tag helper
- Make simple controller to render view
- Create a CS content-system page, using the option to override the content system itself and just point direct to a URL, such that the link still shows up on the menu
- For that latter point, it looks like cs allows the visibility of that nav link to be controlled by Role (not by Auth Policy)

I can now work on the view and controller etc in the usual ways.

**Comment by CrispinF** [comment]
Thanks @JimKerslake and just for info for all, on the final two points, this can equally be done by adding the page to navigation.xml.
The small advantage of doing it via the CMS is that the pages can be anywhere amongst other content pages. The advantage of navigation.xml is that it is _deployed_ therefore cannot be forgotten, while the CMS approach has to be done by a human.

**Comment by joerichardsESDM** [comment]
> @joerichardsESDM ok I think you are saying that the "SBI" prefix is not part of the identifier, and the applicants & RPA only use the 9 digits? Correct?

Yes the API call we make is in the form SBI=123456789

**Comment by JimKerslake** [comment]
@CrispinF  Another possible advantage of doing it in navigation.xml - 
the visibility of menu items looks to be controllable via dynamic auth policy that way 
(whereas it seems to be Role when done via content pages).

**Comment by CrispinF** [comment]
@JimKerslake see Joe's comments about the SBI... I was wrong earlier when I said it included the letters; it want to be just 9 digits.

And yes you're right about the menu item control in navigation.xml. However I'm not sure that matters in ELM, as it's so simple. Authenticated users are in effect applicants, and the only additional privilege will be a HEFER admin role that gets to see another page listing everything. 
We should still use Auth Policies to control the code, but the menu items could adequately be controlled either way.

And in this site we may well want the functional pages to be in amongst content pages (e.g. Home and Help at either end), so I'd lean towards doing it the way you did the Download page.

**Comment by JimKerslake** [comment]
Well it took ages to figure out how to perform client-side validation on this input - 
I'm using a regex pattern annotation on a C# viewmodel property (the 'correct' way) - 
but I needed to pass in a custom error message (from my textpack resolver) to the validation check 
(but the latter can't be a property of an annotation on a viewmodel, because it's dynamically resolved in the view).

You can override the data-* attributes that would otherwise be inferred from the model annotations:

`<input asp-for="SBI" class="form-control" data-val-regex="@tx.InvalidFormatMessage" />`

Also needed to remember to include the correct reference to the Unobtrusive js scripts (not there by default).

**Comment by CrispinF** [comment]
Auth Policy for controller action (on this and download): HeferRequestPolicy

**Comment by JimKerslake** [comment]
@joerichardsESDM   Needing to wire up your data service into the main web app - assume that's OK and you didn't want them separate for any reason. 

```
     // Register ELM data services
            services.AddElmDataServices();
```

**Comment by joerichardsESDM** [comment]
> @joerichardsESDM Needing to wire up your data service into the main web app - assume that's OK and you didn't want them separate for any reason.
> 
> ```
>      // Register ELM data services
>             services.AddElmDataServices();
> ```

No, that's what I assumed would happen, that is why I spilt them out into a project that could be shared.

**Comment by CrispinF** [comment]
@joerichardsESDM please change menu text to "Request".

**Comment by JimKerslake** [comment]
@CrispinF   At present the URLs for these two HEFER pages are  /requesthefer and /downloadhefer  though either can be changed easily - just on the controller 'Route' annotations (and the target property of the CS content pages) 

You'll probably want some attention to what happens when this request hits the server - I'm just using Joe's service to create and save the entity... and I'm sure you'll want more pre-checking etc.  
(Note to self - add logging too).

Also you'll want to think about the routing-back to a 'confirmed' page... I just put a very basic view in for that.  I can easily make the text of that page custom-editable too, by adding it to the same textpack class.  Maybe two such, one for success, one for failure...?

**Comment by CrispinF** [comment]
@JimKerslake the URLs are fine, though personally I would always include a hyphen between the "words" for legibility (do that if it is truly quick and simple). It's the menu text I suggested changing for this and the Download page, to just "Request" and "Download".

Not sure what you mean by the second para. Discuss?

I think there's only a success/confirmed page - this part doesn't yet know about any failures down the line.
For simplicity this could even be a content page with pre-defined URL (configured to now show in menu/sitemap). Then you don't need to do any work, and client can edit the page to heart's content.
e.g. /request-hefer-submitted

**Comment by CrispinF** [comment]
@JimKerslake yes logging of request submission please, including the SBI and user guid (no username or email).

**Comment by JimKerslake** [comment]
The logger config had been significantly de-tuned to reject anything below 'warning' - 
I have re-set it to care about 'information' log events too.

**Comment by CrispinF** [comment]
@JimKerslake I see after submission we arrive at this page: https://elm-hefer-staging.esdm.co.uk/request-hefer/createrequest
How is the content of that page managed?

Very minor layout tweak: could this label be full width without line breaking?
![image](https://user-images.githubusercontent.com/10574730/116377406-0f4feb80-a809-11eb-941b-606b79725fc9.png)

**Comment by JimKerslake** [comment]
@CrispinF content of that page is not currently managed - it's just a view - I could add its text and header to the text pack, as I was suggesting above, or as you say just rely on 'special knowledge' that a cs content page will exist on a particular url.   

Was thinking the former gives more flexibility if you want to pass a model back to the confirmation page, because I can imagine that page needing to handle various failure scenarios too... I can think of a few.

**Comment by CrispinF** [comment]
@JimKerslake agreed, please add an editable H1 title and an editable text block (html).

**Comment by CrispinF** [comment]
@sarahpoppy this page is functional now, and you can edit the labels too. The page appears only once signed in.
https://hefer-data.org.uk/request-hefer
(likewise on staging).

Please test and verify that you can edit the texts.

Ability to edit labels is controlled by membership of a role, so once they are done we can take you out of that role and you will no longer see little "edit" links.

You will soon be able to edit the texts on the page you see after submissions.

**Comment by JimKerslake** [comment]
Interesting small issue with the in situ editor - (general issue, not in any way ELM-specific)

when a view is returned from a POST controller action you can certainly render edit links on it, 
but when using the editor, on posting any text changes back, 
it attempts to re-load the page to show those changes - 
which can only really be by GET (You would not want to consider re-POSTing the form again).   
So the editing experience breaks unless you implement a GET for the same view, just for this editing process.

**Comment by sarahpoppy** [comment]
I don't appear to have any means of editing any pages on either live or staging, and get a access denied message on the request HEFER page @CrispinF @JimKerslake

**Comment by CrispinF** [comment]
@joerichardsESDM (or perhaps @JimKerslake ?) - something wrong on this page now on staging. It's set to be visible to any auth user:
![image](https://user-images.githubusercontent.com/10574730/116665610-e444d380-a991-11eb-8b77-c308b8330b9b.png)

But when an auth user (who is in no other role) visits the page, we get 
![image](https://user-images.githubusercontent.com/10574730/116665674-f4f54980-a991-11eb-86eb-e93bcb978e74.png)

But nothing in the log, from what I can see.

**Comment by CrispinF** [comment]
@sarahpoppy stuff went wrong in this update, we'll let you know when ready for test again.

**Comment by CrispinF** [comment]
@sarahpoppy OK, you should now be able to edit texts on the Request page on both staging and production. 
On staging you can also edit the texts on the page you land on after submitting. This is not yet deployed to production.
